src/bin/pg_autoctl/keeper.c

```diff
@@ -1165,6 +1165,30 @@ keeper_register_and_init(Keeper *keeper, NodeState initialState)
 	/* use a special connection retry policy for initialisation */
 	(void) pgsql_set_init_retry_policy(&(keeper->monitor.pgsql));
 
+	/*
+	 * When registering to the monitor, we get assigned a nodeId, that we keep
+	 * preciously in our state file. We need to have a local version of the
+	 * nodeId that is the same one as on the monitor.
+	 *
+	 * In particular, if we fail to update our local state file, we should
+	 * cancel our registration, because there's no way we can re-discover our
+	 * nodeId later.
+	 *
+	 * We register to the monitor in a SQL transaction that we only COMMIT
+	 * after we have updated our local state file. If we fail to do so, we
+	 * ROLLBACK the transaction, and thus we are not registered to the monitor
+	 * and may try again. If we are disconnected halfway through the
+	 * registration (process killed, crash, etc), then the server issues a
+	 * ROLLBACK for us upon disconnection.
+	 */
+	if (!pgsql_execute(&(monitor->pgsql), "BEGIN"))
+	{
+		log_error("Failed to open a SQL transaction to register this node");
+
+		unlink_file(config->pathnames.state);
+		return false;
+	}
+
 	if (!monitor_register_node(monitor,
```

Comment:
I think this error case should also `goto rollback`.

Reply:
Yeah that's a fair point, done that way now.